mlir/lib/Gradient/Transforms/vector_lowering.cpp

```diff
@@ -0,0 +1,231 @@
+
+        Value predicate =
+            b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::eq, sizeVal, capacityVal);
+        b.create<scf::IfOp>(loc, predicate, [&](OpBuilder &thenBuilder, Location loc) {
+            Value two = thenBuilder.create<arith::ConstantIndexOp>(loc, 2);
+            Value newCapacity = thenBuilder.create<arith::MulIOp>(loc, capacityVal, two);
+            Value oldElements = thenBuilder.create<memref::LoadOp>(loc, elementsField);
+            Value newElements = thenBuilder.create<memref::ReallocOp>(
+                loc, cast<MemRefType>(oldElements.getType()), oldElements, newCapacity);
+            thenBuilder.create<memref::StoreOp>(loc, newElements, elementsField);
+            thenBuilder.create<memref::StoreOp>(loc, newCapacity, capacityField);
+            thenBuilder.create<scf::YieldOp>(loc);
+        });
+
+        Value elementsVal = b.create<memref::LoadOp>(loc, elementsField);
+        b.create<memref::StoreOp>(loc, value, elementsVal,
+                                  /*indices=*/sizeVal);
+
+        Value one = b.create<arith::ConstantIndexOp>(loc, 1);
+        Value newSize = b.create<arith::AddIOp>(loc, sizeVal, one);
+
+        b.create<memref::StoreOp>(loc, newSize, sizeField);
+        b.create<func::ReturnOp>(loc);
+        return SymbolRefAttr::get(ctx, funcName);
+    }
+
+    void emitPush(Location loc, Value value, OpBuilder &b, FlatSymbolRefAttr pushFn) const
+    {
+        b.create<func::CallOp>(loc, pushFn, /*results=*/TypeRange{},
+                               /*operands=*/ValueRange{dataField, sizeField, capacityField, value});
+    }
+};
+
+struct LowerInitVector : public OpConversionPattern<VectorInitOp> {
+    using OpConversionPattern<VectorInitOp>::OpConversionPattern;
+
+    LogicalResult matchAndRewrite(VectorInitOp op, OpAdaptor adaptor,
+                                  ConversionPatternRewriter &rewriter) const override
+    {
+        SmallVector<Type> resultTypes;
+        if (failed(getTypeConverter()->convertType(op.getType(), resultTypes))) {
+            op.emitError() << "Failed to convert type " << op.getType();
+            return failure();
+        }
+        Value capacity = rewriter.create<arith::ConstantIndexOp>(op.getLoc(), 1);
+        Value initialSize = rewriter.create<arith::ConstantIndexOp>(op.getLoc(), 0);
+        auto dataType = resultTypes[0].cast<MemRefType>();
+        auto sizeType = resultTypes[1].cast<MemRefType>();
+        auto capacityType = resultTypes[2].cast<MemRefType>();
+        Value buffer = rewriter.create<memref::AllocOp>(
+            op.getLoc(), dataType.getElementType().cast<MemRefType>(),
+            /*dynamicSize=*/capacity);
+        Value bufferField = rewriter.create<memref::AllocaOp>(op.getLoc(), dataType);
+        Value sizeField = rewriter.create<memref::AllocaOp>(op.getLoc(), sizeType);
+        Value capacityField = rewriter.create<memref::AllocaOp>(op.getLoc(), capacityType);
```

Comment:
Just double checking the use of alloca here, couldn't these values go out scope while the vector is still in use?

Reply:
Hm, there is this possibility if the vector is returned from a function. I can make these `alloc` instead and hope that the `-promote-buffers-to-stack` promotes them when it is valid to do so.